sklearn/neighbors/lof.py

```diff
@@ -0,0 +1,315 @@
+        opposite of the outlier_factor_ attribute.
+
+        Parameters
+        ----------
+        X : array-like, shape (n_samples, n_features)
+            The query sample or samples to compute the Local Outlier Factor
+            w.r.t. the training samples.
+
+        Returns
+        -------
+        lof_scores : array, shape (n_samples,)
+            The Local Outlier Factor of each input samples. The lower,
+            the more abnormal.
+        """
+        check_is_fitted(self, ["threshold_", "outlier_factor_",
+                               "_distances_fit_X_",
+                               "_neighbors_indices_fit_X_"])
+
+        X = check_array(X, accept_sparse='csr')
+
+        distances_X, neighbors_indices_X = self.kneighbors(X)
+        # as bigger is better:
+        return -self._local_outlier_factor(distances_X, neighbors_indices_X)
+
+    def _local_outlier_factor(self, distances_X, neighbors_indices_X):
+        """Compute the local outlier factor (LOF)
+
+        It is the average of the ratio of the local reachability density of
+        a sample and those of its k-nearest neighbors.
+
+        Parameters
+        ----------
+        distances_X : array, shape (n_query, self.n_neighbors)
+            Distances to the neighbors (in the training samples self._fit_X) of
+            each query point to compute the LRD.
+
+        neighbors_indices : array, shape (n_query, self.n_neighbors)
+            Neighbors indices (of each query point) among training samples
+            self._fit_X.
+
+        Returns
+        -------
+        lof : numpy array, shape (n_samples,)
+            The LOF of X. The lower, the more normal.
+        """
+        n_samples = distances_X.shape[0]
+
+        # Compute the local_reachability_density of samples X:
+        X_lrd = self._local_reachability_density(distances_X,
+                                                 neighbors_indices_X)
+        lrd_ratios_array = np.zeros((n_samples, self.n_neighbors))
+
+        # Avoid re-computing X_lrd if same parameters:
+        if not (np.all(distances_X == self._distances_fit_X_) *
+                np.all(self._neighbors_indices_fit_X_ == neighbors_indices_X)):
+            lrd = self._local_reachability_density(
+                self._distances_fit_X_, self._neighbors_indices_fit_X_)
+        else:
+            lrd = X_lrd
+
+        lrd_ratios_array = lrd[neighbors_indices_X] / X_lrd[:, np.newaxis]
+
+        return np.mean(lrd_ratios_array, axis=1)
+
+    def _local_reachability_density(self, distances_X, neighbors_indices):
+        """The local reachability density (LRD)
+
+        The LRD of a sample is the inverse of the average reachability
+        distance of its k-nearest neighbors.
+
+        Parameters
+        ----------
+        distances_X : array, shape (n_query, self.n_neighbors)
+            Distances to the neighbors (in the training samples self._fit_X) of
```

Comment:
I meant putting backticks around ``self._fit_X``. a) for nicer highlighting b) I'm not sure sphinx will render the current version correctly because of the underscore. But I might be paranoid.